File: app/src/main/java/com/example/mlallemant/mentalbattle/Utils/Calculation.java
package com.example.mlallemant.mentalbattle.Utils;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Random;

/**
 * Created by m.lallemant on 19/10/2017.
 */

public class Calculation implements Parcelable {

    private static final int MAX_VALUE_ADDITION = 150;
    private static final int MAX_RANGE = 100;

    private Integer result;
    private final String calculText;

    public Calculation() {
        Integer A, B;

        A = getRandomNumber();
        B = getRandomNumber();
        result = calculResult(A, B);
        while (result > MAX_VALUE_ADDITION && result < -MAX_VALUE_ADDITION) {
            A = getRandomNumber();
            B = getRandomNumber();
            result = calculResult(A, B);
        }

        calculText = makeString(A, B);
    }

    private Integer getRandomNumber() {
        final Integer a;

        final Random random = new Random();
        a = (random.nextInt(MAX_RANGE + 1 + MAX_RANGE) - MAX_RANGE);

        return a;
    }

    private Integer calculResult(final Integer a, final Integer b) {
        return a + b;
    }

    private String makeString(final Integer a, Integer b) {
        final String text;
        final String sign;

        if (b < 0) {
            sign = "-";
            b *= -1;
        } else {
            sign = "+";
        }

        text = a + " " + sign + " " + b;
        return text;
    }

    public Integer getResult() {
        return result;
    }

    public String getCalculText() {
        return calculText;
    }

    protected Calculation(final Parcel in) {
        result = in.readByte() == 0x00 ? null : in.readInt();
        calculText = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        if (result == null) {
            dest.writeByte((byte) (0x00));
        } else {
            dest.writeByte((byte) (0x01));
            dest.writeInt(result);
        }
        dest.writeString(calculText);
    }

    @SuppressWarnings("unused")
    public static final Parcelable.Creator<Calculation> CREATOR = new Parcelable.Creator<Calculation>() {
        @Override
        public Calculation createFromParcel(final Parcel in) {
            return new Calculation(in);
        }

        @Override
        public Calculation[] newArray(final int size) {
            return new Calculation[size];
        }
    };
}

File: app/src/main/java/com/example/mlallemant/mentalbattle/utils/Calculation.kt
package com.example.mlallemant.mentalbattle.utils

import android.os.Parcel
import android.os.Parcelable
import java.util.*

/**
 * Created by m.lallemant on 19/10/2017.
 */
class Calculation : Parcelable {
    var result: Int?
        private set
    val calculText: String?

    constructor() {
        var A: Int
        var B: Int
        A = randomNumber
        B = randomNumber
        result = calculResult(A, B)
        while (result!! > MAX_VALUE_ADDITION && result!! < -MAX_VALUE_ADDITION) {
            A = randomNumber
            B = randomNumber
            result = calculResult(A, B)
        }
        calculText = makeString(A, B)
    }

    private val randomNumber: Int
        private get() {
            val a: Int
            val random = Random()
            a =
                random.nextInt(MAX_RANGE + 1 + MAX_RANGE) - MAX_RANGE
            return a
        }

    private fun calculResult(a: Int, b: Int): Int {
        return a + b
    }

    private fun makeString(a: Int, b: Int): String {
        var b = b
        val text: String
        val sign: String
        if (b < 0) {
            sign = "-"
            b *= -1
        } else {
            sign = "+"
        }
        text = "$a $sign $b"
        return text
    }

    protected constructor(`in`: Parcel) {
        result = if (`in`.readByte().toInt() == 0x00) null else `in`.readInt()
        calculText = `in`.readString()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        if (result == null) {
            dest.writeByte(0x00.toByte())
        } else {
            dest.writeByte(0x01.toByte())
            dest.writeInt(result!!)
        }
        dest.writeString(calculText)
    }

    companion object CREATOR : Parcelable.Creator<Calculation> {
        private const val MAX_VALUE_ADDITION = 150
        private const val MAX_RANGE = 100

        override fun createFromParcel(parcel: Parcel): Calculation {
            return Calculation(parcel)
        }

        override fun newArray(size: Int): Array<Calculation?> {
            return arrayOfNulls(size)
        }
    }
}